crates/typst-library/src/text/shaping.rs

```diff
@@ -488,17 +488,24 @@ impl<'a> ShapedText<'a> {
         }
 
         // Find any glyph with the text index.
-        let mut idx = self
-            .glyphs
-            .binary_search_by(|g| {
-                let ordering = g.range.start.cmp(&text_index);
-                if ltr {
-                    ordering
-                } else {
-                    ordering.reverse()
-                }
-            })
-            .ok()?;
+        let cmp = |g: &ShapedGlyph| {
+            let ordering = g.range.start.cmp(&text_index);
+            if ltr {
+                ordering
+            } else {
+                ordering.reverse()
+            }
+        };
+        let mut idx = match self.glyphs.binary_search_by(cmp) {
+            Ok(idx) => idx,
+            Err(idx) => {
+                // Handle the special case where we break before a '\n'
+                return (idx > 0
+                    && self.glyphs[idx - 1].range.end == text_index
+                    && self.text[text_index - self.base..].starts_with('\n'))
+                .then_some(idx);
```

Comment:
could we replace this check by this one?

```rust
return (idx == 0 || idx == self.glyphs.len()).then_some(idx);
```

from my understanding, the problem is that `self.text` covers more text than `self.glyphs` because of codepoints that don't result in glyphs. however, there might be other cases of this than '\n'. if the binary search says "insert at 0 or len", that means that the target text index exceeds the shaped glyphs, i.e. the `text_range` passed to `find_safe_to_break` is at least as large as the one covered by shaped glyphs. at the same time, it is at most as large as the ones covered by the shaped glyphs, since it's a subslice from an even larger range that produced the shaped glyphs. so we should be fine with just returning the full range, right?

this wouldn't handle the case where there are invisible glyphs _within_ the shaped text (not at the bounds of it), but it would not be safe to assume that we can break there anyway, I think.

Reply:
> this wouldn't handle the case where there are invisible glyphs

This is exactly what I want to handle. Consider the following text:

`字\n字`

I want the slice that consists of the first character `字` to be considered safe. `\n` will always be safe to break, and as you pointed out other characters may be not safe so I filter them out.

Reply:
Ah, I see now! I somehow thought that the `\n` will always be at the end of the text. I think this code path would benefit from a longer comment explaining things.

And the first part `self.glyphs[idx - 1].range.end == text_index` is necessary because there might be even more invisible glyphs before the '\n'?

Reply:
Technically, I think we're using harfbuzz/rustybuzz in an uncanonical way here by giving it the full string with newlines, but so far it seemed to work out fine ...